src/frontend/src/binder/expr/function.rs

```diff
@@ -252,80 +188,203 @@ impl Binder {
+        // Syntax:
+        // aggregate_name (expression [ , ... ] [ order_by_clause ] ) [ FILTER ( WHERE
+        //   filter_clause ) ]
+        // aggregate_name (ALL expression [ , ... ] [ order_by_clause ] ) [ FILTER ( WHERE
+        //   filter_clause ) ]
+        // aggregate_name (DISTINCT expression [ , ... ] [ order_by_clause ] ) [ FILTER ( WHERE
+        //   filter_clause ) ]
+        // aggregate_name ( * ) [ FILTER ( WHERE filter_clause ) ]
+
+        assert!(!matches!(kind, agg_kinds::ordered_set!()));
+
+        if f.within_group.is_some() {
+            return Err(ErrorCode::InvalidInputSyntax(format!(
+                "WITHIN GROUP is not allowed for non-ordered-set aggregation `{}`",
+                kind
+            ))
+            .into());
+        }
+
+        let args: Vec<_> = f
+            .args
+            .iter()
+            .map(|arg| self.bind_function_arg(arg.clone()))
+            .flatten_ok()
+            .try_collect()?;
+        let order_by = OrderBy::new(
+            f.order_by
+                .into_iter()
+                .map(|e| self.bind_order_by_expr(e))
+                .try_collect()?,
+        );
+
+        if f.distinct {
+            if kind == AggKind::ApproxCountDistinct {
+                return Err(ErrorCode::InvalidInputSyntax(format!(
+                    "DISTINCT is not allowed for approximate aggregation `{}`",
+                    kind
+                ))
+                .into());
+            }
+
+            if args.is_empty() {
+                return Err(ErrorCode::InvalidInputSyntax(format!(
+                    "DISTINCT is not allowed for aggregate function `{}` without args",
+                    kind
+                ))
+                .into());
+            }
+
+            // restrict arguments[1..] to be constant because we don't support multiple distinct key
+            // indices for now
+            if args.iter().skip(1).any(|arg| arg.as_literal().is_none()) {
+                return Err(ErrorCode::NotImplemented(
+                    "non-constant arguments other than the first one for DISTINCT aggregation is not supported now"
+                        .to_string(),
+                    None.into(),
+                )
+                .into());
+            }
+
+            // restrict ORDER BY to align with PG, which says:
+            // > If DISTINCT is specified in addition to an order_by_clause, then all the ORDER BY
+            // > expressions must match regular arguments of the aggregate; that is, you cannot sort
+            // > on an expression that is not included in the DISTINCT list.
+            if !order_by.sort_exprs.iter().all(|e| args.contains(&e.expr)) {
+                return Err(ErrorCode::InvalidInputSyntax(format!(
+                    "ORDER BY expressions must match regular arguments of the aggregate for `{}` when DISTINCT is provided",
+                    kind
+                ))
+                .into());
+            }
+        }
```

Comment:
Actually we can _execute_ the cases where ORDER BY columns not in argument list, but the semantics may be ambiguous as mentioned by @st1page